src/Compilers/Core/Portable/InternalUtilities/StackGuard.cs

```diff
@@ -29,5 +30,27 @@ public static void EnsureSufficientExecutionStack(int recursionDepth)
                 RuntimeHelpers.EnsureSufficientExecutionStack();
             }
         }
+
+        internal static bool TryEnsureSufficientExecutionStack(int recursionDepth, bool throwOnFailure)
+        {
+            try
+            {
+                EnsureSufficientExecutionStack(recursionDepth);
+                return true;
+            }
+            catch (InsufficientExecutionStackException) when (!throwOnFailure)
+            {
+                return false;
+            }
+        }
+
+        internal static TResult ExecuteOnNewExecutionStack<TResult>(Func<TResult> execute)
+        {
+            var task = Task.Run(() => execute());
+            // Wait on the task without inlining the task on this thread.
+            Task.WhenAny(task).Wait();
+            // Return result, propagating any exception.
+            return task.GetAwaiter().GetResult();
+        }
```

Comment:
Another general question about the change: if parsing is not longer a bottleneck, then other areas will be. Is the plan that we apply a similar technique to those as well? 
If so, I'd assume that the same option would drive the new behavior as well, is that correct?

Reply:
Also, when/where are *we* going to set this flag? This PR didn't expose it to command-line or MSBuild. What's the plan?

Reply:
Yes we would need to apply a similar technique to phases after parsing to ensure end-to-end compilation scenarios work with deeply nested structures.